ci/infra/libvirt/master-instance.tf

```diff
@@ -126,6 +126,7 @@ resource "null_resource" "master_reboot" {
     command = <<EOT
 ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null $user@$host sudo reboot || :
 # wait for ssh ready after reboot
+until nc -zv $host 22; do sleep 5; done
 ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -oConnectionAttempts=60 $user@$host /usr/bin/true
```

Comment:
so if the problem is that we want to avoid continuing while passwordless login does not yet succeed, then we should do this here rather than hacking around with nc. you can simply add

```
-o PasswordAuthentication=no -o ChallengeResponseAuthentication=no 
```

Reply:
The problem is that it should use the keys to authenticate via the ssh-agent instead of the password. However, right after the sshd server starts after the booting, there is a very small period of time in which ssh-agent fails to communicate the key. If the ssh request is done over that period of time, then Terraform concludes that the task has failed because it was not able to authenticate.

Using a tool (in this case netcat) to check that the ssh port is up and running before trying to do ssh seems to me like best engineering practices and no hack. TBH, perhaps I missed it in the documentation but I am surprised that Terraform does not include any syntax to check if ports are up and running, to me that is basic functionality. Anyway, not doing any ssh until we are sure that the port is up and running avoids the problem I have just described, that's why I consider it as a good solution

Reply:
I understand, so ssh will not return the exit code of "/usr/bin/true" (which is 0) but some other error code which indicates "ssh has failed to run the command". and that leads to terraform treating that as a failure. 

Your command will first check whether the port is listening to requests, and then try the request. if I understand the problem properly, then the race is still there, its just much smaller ,because the "restart" might not be finished after the nc succeeded but the ssh attempt afterwards is started. thats why I said "hacking around" (I didn"t say your code is a hack btw, I meant it in the sense of "workaround"). 


we could simply retry the ssh command until it exits with an exit code of 0 (or give up after like 3-5 attempts. that would close the race if I understand your problem description correctly, because then we can be confident that further ssh attempts will succeed.

Reply:
Another thing is that there might be some sleeping missing waiting for the node shutdown (because we're just executing a reboot command) so chances are there that we're still hitting the instance before the reboot, rather than waiting until reboot finished with the new ssh being up.

Reply:
Yes, that's a good point. @jcaamano was looking at it yesterday

Reply:
> I understand, so ssh will not return the exit code of "/usr/bin/true" (which is 0) but some other error code which indicates "ssh has failed to run the command". and that leads to terraform treating that as a failure.
> 
> Your command will first check whether the port is listening to requests, and then try the request. if I understand the problem properly, then the race is still there, its just much smaller ,because the "restart" might not be finished after the nc succeeded but the ssh attempt afterwards is started. thats why I said "hacking around" (I didn"t say your code is a hack btw, I meant it in the sense of "workaround").
> 
> we could simply retry the ssh command until it exits with an exit code of 0 (or give up after like 3-5 attempts. that would close the race if I understand your problem description correctly, because then we can be confident that further ssh attempts will succeed.
> 
> that latter approach is for example what the OpenStack Cloud team used to wait for VMs to be finished booting in order to ssh them. I think it might be worthwhile and its definitely proven in practice.

Not exactly. The command which is executed is:

`ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -oConnectionAttempts=60 $user@$host /usr/bin/true`

That one tries 60 times. That is fine as long as the receiving error is, no IP:
`ssh: connect to host 10.17.3.0 port 22: No route to host`

Or port 22 not up:
`ssh: connect to host 10.17.3.0 port 22: Connection refused`

But it is not fine and terraform fails when there is an authentication error due to the ssh-agent problem `sign_and_send_pubkey: signing failed: communication with agent failed`. Terraform fails regardless of how many times it has tried before, it could be the first time it tries and it will not try another 59 times, it will simply fail.

You are right that there is a race condition (the results in the ssh-agent problem) and this solution is not removing it but reducing the likelihood of it. I actually haven't seen it again since I started using this patch but there is probably still a chance. Therefore, we can consider other options but it cannot be by using ssh because we will be hitting the ssh-agent problem again. One alternative that would probably work is to disable ssh-agent in that terraform task but as ssh-agent is used across our Terraform deployment, I expect a lot of negative reviews.

Having said so (sorry for extending myself so much ;) ), I still think that checking the ssh port is up and running before doing ssh is something that should be always done

Reply:
I think I understand. Terraform fails when there is a comnabd exiting with exit code nonzero. But you can capture that and sleep while the exit code is nonzero like (pseudo code ahead)


```
while ! ssh ... ; do
  sleep 1
  if retriedtoooften; then
     exit 1
  fi
done
```